openedx/core/djangoapps/user_api/preferences/tests/test_views.py

```diff
@@ -0,0 +1,392 @@
+        self.send_get(self.different_client, expected_status=404)
+        self.send_put(self.different_client, "new_value", expected_status=404)
+        self.send_delete(self.different_client, expected_status=404)
+
+    @ddt.data(
+        ("client", "user"),
+        ("staff_client", "staff_user"),
+    )
+    @ddt.unpack
+    def test_get_unknown_user(self, api_client, username):
+        """
+        Test that requesting a user who does not exist returns a 404.
+        """
+        client = self.login_client(api_client, username)
+        response = client.get(
+            reverse(self.url_endpoint_name, kwargs={'username': "does_not_exist", 'preference_key': self.test_pref_key})
+        )
+        self.assertEqual(404, response.status_code)
+
+    def test_get_preference_does_not_exist(self):
+        """
+        Test that a 404 is returned if the user does not have a preference with the given preference_key.
+        """
+        self._set_url("does_not_exist")
+        self.client.login(username=self.user.username, password=self.test_password)
+        response = self.send_get(self.client, expected_status=404)
+        self.assertIsNone(response.data)
+
+    @ddt.data(
+        ("client", "user"),
+        ("staff_client", "staff_user"),
+    )
+    @ddt.unpack
+    def test_get_preference(self, api_client, user):
+        """
+        Test that a client (logged in) can get her own preferences information. Also verifies that a "is_staff"
+        user can get the preferences information for other users.
+        """
+        client = self.login_client(api_client, user)
+        response = self.send_get(client)
+        self.assertEqual(self.test_pref_value, response.data)
+
+        # Test a different value.
+        UserPreference.set_preference(self.user, "dict_pref", {"int_key": 10})
+        self._set_url("dict_pref")
+        response = self.send_get(client)
+        self.assertEqual("{'int_key': 10}", response.data)
+
+    def test_create_preference(self):
+        """
+        Test that a client (logged in) can create a preference.
+        """
+        self.client.login(username=self.user.username, password=self.test_password)
+
+        # Verify that a new preference can be created
+        self.send_delete(self.client)
+        new_value = "new value"
+        self.send_put(self.client, new_value)
+        response = self.send_get(self.client)
+        self.assertEqual(new_value, response.data)
+
+    def test_create_preference_too_long_key(self):
+        """
+        Test that a client cannot create preferences with bad keys
+        """
+        self.client.login(username=self.user.username, password=self.test_password)
+
+        too_long_preference_key = "x" * 256
+        new_value = "new value"
+        self._set_url(too_long_preference_key)
+        response = self.send_put(self.client, new_value, expected_status=400)
+        self.assertEqual(new_value, response.data)
```

Comment:
That was bad cut-and-paste, and it was always failing on the line before. I've fixed it.